GoogleForCreators/web-stories-wp
# Workspace: Keyboard navigation in Grid View

## Summary
Add keyboard navigation support to GridView

## Relevant Technical Choices

The main change is the addition of the `useGridViewKeys` custom hook.  It takes inspiration and patterns from the `useCarouselKeys` hook.  


## To-do
- Add tests for up/down arrow key navigation/arrangement. See https://github.com/puppeteer/puppeteer/issues/6195


## User-facing changes
Users should now be able to navigate in the grid view using their keyboard.

## Testing Instructions
Launch into the editor's grid view and test the following scenarios:

1. Use tab to jump between "Back" button, "Preview size" range control (three stops), and the page list.
2. When jumping to the page list the currently selected page should be focused instead. Not the whole list.
3. Use ↕️ and ↔️ keys to navigate through the grid horizontally and vertically, and with wrapping.
4. Use mod+↕️ and mod+↔️ to reorder pages.
5. Use Enter to select a page.
6. The focus should be trapped inside the dialog. E.g. if at any point the focus leaves the dialog while it's still open, the focus should be returned back to the dialog.
7. Use Esc to exit the dialog at any time.

Also validate that the `edit-story` karma tests pass for this build as integrations tests were added for the above 7 requirements.

---



#2717
